src/modules/commander/Commander.cpp

```diff
@@ -1781,9 +1782,27 @@ Commander::run()
 						tune_mission_ok(true);
 					}
 				}
+
+				if (mission_result.finished) {
+					_last_mission_result_finished = mission_result.timestamp;
+				}
 			}
 		}
 
+		// When auto mission/rtl/land is finished, transition back to the most recently-used manual mode.
+		// This check is done using time because we cannot guarantee that the disarm will happen first or
+		// the mission_result message will come in first.
+		const bool disarmed_and_mission_finished = !armed.armed
+				&& hrt_elapsed_time(&last_disarmed_timestamp) < MISSION_FINISH_DISARM_TIMEOUT
+				&& hrt_elapsed_time(&_last_mission_result_finished) < MISSION_FINISH_DISARM_TIMEOUT;
+
+		if (disarmed_and_mission_finished && is_auto_state(internal_state.main_state)
+		    && last_non_auto_state != commander_state_s::MAIN_STATE_MAX) {
+			// This branch will only happen once per mission finish because, after transitioning to the non-auto
+			// state, is_auto_state(internal_state.main_state) will be false.
```

Comment:
If a user switches to mission again within `MISSION_FINISH_DISARM_TIMEOUT`, it will be true again, so the comment is not quite right.
Since the timeout is 500ms, the logic should be fine, but I'd be more comfortable if multiple triggering of this is prevented (for example we might have to increase `MISSION_FINISH_DISARM_TIMEOUT` for some reason).

How does `COM_DISARM_LAND` matter for this?

Reply:
> How does `COM_DISARM_LAND` matter for this?

I think it won't work if `COM_DISARM_LAND` is bigger, right? Because too much time has lapsed by the time it disarms.

Reply:
I changed this to not rely on timing. Now, it just makes sure all of the following conditions are met:

 - The vehicle just now disarmed
 - The vehicle is currently landed
 - The current mode is an auto mode
 - The current mission is finished

Here's the code:
```
bool just_disarmed = !armed.armed && was_armed;
		bool just_finished_auto_mission = is_auto_state(internal_state.main_state) && _mission_result_sub.get().finished;
		bool last_state_valid = last_non_auto_state != commander_state_s::MAIN_STATE_MAX;

		if(just_disarmed && land_detector.landed && just_finished_auto_mission && last_state_valid){
			PX4_INFO("Just finished auto mission, transitioning back to last manual mode.");
			main_state_transition(status, last_non_auto_state, status_flags, &internal_state);
		}
```
Is this sufficient? Or did I miss some edge case?